model/flow/identity.go

```diff
@@ -113,11 +123,20 @@ func encodableFromIdentity(iy Identity) (encodableIdentity, error) {
 }
 
 func (iy Identity) MarshalJSON() ([]byte, error) {
+	var identity interface{}
 	encodable, err := encodableFromIdentity(iy)
 	if err != nil {
 		return nil, fmt.Errorf("could not convert identity to encodable: %w", err)
 	}
-	data, err := json.Marshal(encodable)
+
+	// if the address is empty, suppress the Address field in the output json
+	if encodable.Address == "" {
```

Comment:
So, at the moment we need this `MarshalJSON` and `StrippedInmemSnapshot` to each work in sync with each other.

- are there known cases where we want to preserve the `Address` field?
- if not, couldn't we simply convert all `identity` fields here to remove the Address in one step?

Reply:
You mean valid cases where the `address` is an empty string but we still want to retain it? I don't think so but will have to check.
In the regular case, the address will be non empty and we do want to retain it.

Reply:
> In the regular case, the address will be non empty and we do want to retain it.

Yep, of course.

What I mean is this: we have two data flows, branching on whether you're producing a file for an unstaked node. One is marshalling as usual, another one is using two sequential components to suppress Addresses.

THose sequential components are:
- replacing Addresses by the empty string,
- detecting the empty string addresses, and for those identities, converting them to a struct which Address is skipped on marshalling.

I'm saying we can:
- branch on whether you're producing a file for an unstaked node,
- in the case we are, convert every identity to a struct which Address field is omitted.

Reply:
The identity structs are used within the different structs that make up the snapshot. e.g. Epoch, Execution Result -> Service Event. Hence, defining a new Identity struct  (e.g. `StealthIdentity`) would also require re-defining these enclosing structs to their new `Stealthxxx` version.

Reply:
Yeah, which is why I linked to a way of doing this dynamically here: https://github.com/onflow/flow-go/pull/1266#pullrequestreview-751736361

This redefines the fields to strip the Address w/o defining a new struct.

Reply:
Oh, I did try that earlier. There are a fews things I ran into -
1. How convert from the new struct type back to an identity type -
This doesn't work :(  - `newIdentity := reflect.ValueOf(newValue).Interface().(flow.Identity)`
2. Tags may still be lost even after the conversion - https://stackoverflow.com/a/32189702
3. The flow.Identity is converted to yet another type - Encodable Identity [when doing the marshalling](https://github.com/onflow/flow-go/blob/master/model/flow/identity.go#L95-#L116). Hence, these tags would have to be copied over to that type.

don't mind investigating this further but I feel I may have hit a wall